vendor/src/github.com/docker/libnetwork/datastore/datastore.go

```diff
@@ -283,11 +305,31 @@ func (ds *datastore) Watch(kvObject KVObject, stopCh <-chan struct{}) (<-chan KV
 				kvoCh <- dstO
 			}
 		}
+
+		// Wait on watch channel for a re-trigger when datastore becomes active
+		<-watchCh
+
+		kvpCh, err = ds.store.Watch(Key(kvObject.Key()...), sCh)
+		if err != nil {
+			log.Printf("Could not watch the key %s in store: %v", Key(kvObject.Key()...), err)
+		}
+
+		goto retry_watch
```

Comment:
I understand the intent, this is a simple barrier to restart the watches when the store is back up (as this is not a `WatchTree` but a set of `Watch`).

I still think that the logic can be greatly improved by reorganizing and splitting this function in two (one for the retry logic and handling the barrier, and the other for the Watch loop). This would be much cleaner and you also get rid of the labels..